offchain/authority-claimer/src/claimer.rs

```diff
@@ -0,0 +1,73 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+use async_trait::async_trait;
+use tracing::{trace, warn};
+
+use crate::{listener::BrokerListener, sender::ClaimSender};
+
+/// The `AuthorityClaimer` starts an event loop that waits for claim messages
+/// from the broker, and then sends the claims to the blockchain.
+///
+/// It uses a `BrokerListener` for listening for messages from the broker.
+///
+/// It also uses a `ClaimSender` that interacts with the blockchain and
+/// effectively submits the claims.
+#[async_trait]
+pub trait AuthorityClaimer<'a, L: BrokerListener, S: ClaimSender>
+where
+    L: 'a + Sync,
+    S: 'a,
+{
+    async fn start(
+        &'a self,
+        broker_listener: L,
+        claim_sender: S,
+    ) -> Result<(), AuthorityClaimerError<S, L>> {
```

Comment:
This construction is curious to me, what's the `self` for? Couldn't this be just a normal function, and the type parameter be part of the function? Or Rust's `impl` feature (e.g. `broker_listener: impl BrokerListener`)?

Reply:
You are a lifesaver, I fought for always with Rust and hated this solution.
I just needed to change stuff to the function...

Reply:
Hahaha, no problem! But don't forget the top review comment on not reading the state of the blockchain anymore. That's the most important.